### ISSUE TITLE
Too many open files error on application server: LiquibaseEntityResolver opens too many file descriptors and never closes them

### ISSUE THREAD

**Issue by blutorange** [liquibase/liquibase#1815]
## Environment
* Java version: OpenJDK 11
* Liquibase Version: newest (4.3.3)
* Liquibase Integration & Version:  Java API (we're using new `new Liquibase().listUnrunChangeSets` manually). It's called from within a webapp with Jetty 9 / Tomcat 8 / Tomcat 9 / JSF 2.3 / Weld 3.1.7.
* Database Vendor & Version: tested with MySQL 8 and PostgreSQL 12, but probably independent of DBMS
* Operating System Type & Version: Ubuntu (tested on 18.04 and 20.10)

## Description
Symptoms are the same as #1449, but not sure if it's the same cause (we're not using Spring). When running Liquibase, it opens many file decriptors and never closes them. They only get closed when the Garbabe  Collector runs, which may be a long in the future. This results in our application server crashing.

I first observed this on a test environment server with Tomcat 8 after upgrading vom liquibase 3 to liquibase 4. The webapp failed to start, and the log contained many `too many open files` messages. Checking via `/proc/pid/fd` indeed revealed that the number of file descriptors had reached the limit. At first I could not reproduce this locally on my machine, but now I can and I think I know why it happens:

![image](https://user-images.githubusercontent.com/2456413/116452065-9464f000-a85d-11eb-9711-c4af435d7033.png)

The problematic part is in [LiquibaseEntitiyResolve](https://github.com/liquibase/liquibase/blob/256675bca43a089e32c940e10ed8ed7a5f9d66f6/liquibase-core/src/main/java/liquibase/parser/core/xml/LiquibaseEntityResolver.java#L38), which is called when Liquibase parses the database changelogs (we've got 11 XML changelog files) after we call `Liquibase#listUnrunChangeSets´.

When the `LiquibaseEntityResolver` is constructed, it [also creates a `fallbackResourceAccessor`](https://github.com/liquibase/liquibase/blob/256675bca43a089e32c940e10ed8ed7a5f9d66f6/liquibase-core/src/main/java/liquibase/parser/core/xml/LiquibaseEntityResolver.java#L19) of type `ClassLoaderResourceAccessor`. The `ClassLoaderResourceAccessor`  loads all JARs of the application, but [does so only lazily](https://github.com/liquibase/liquibase/blob/256675bca43a089e32c940e10ed8ed7a5f9d66f6/liquibase-core/src/main/java/liquibase/resource/ClassLoaderResourceAccessor.java#L51) only when its `#openStreams` method is called. 

Next the `LiquibaseEntityResolver` [creates a `CompositeResourceAccessor`](https://github.com/liquibase/liquibase/blob/256675bca43a089e32c940e10ed8ed7a5f9d66f6/liquibase-core/src/main/java/liquibase/parser/core/xml/LiquibaseEntityResolver.java#L38) with the resource loader from the current scope and the fallback resource loader.  Now, and that's where the error starts happening, the composite resource accessor [always calls `openStreams`](https://github.com/liquibase/liquibase/blob/256675bca43a089e32c940e10ed8ed7a5f9d66f6/liquibase-core/src/main/java/liquibase/resource/CompositeResourceAccessor.java#L34) on _all_ accessors, regardless of whether an earlier found something or not.

This means that `init` method of the fallback `ClassLoaderResourceAccessor` is always called, loading all JARs.

In our case the `LiquibaseEntityResolver` is created about ~20 times, and each time it loads ~300 JARs. The fiel descriptor limit on the server is 4000, so it crashes before liquibase is done.

What I had been wondering is why I cannot observe this behavior locally, but now I've got an explanation for that. The same process I described above _actually does happen_  on my machine, but: the file descriptors are closed implicitly by the `finalize`  of the `ZipFileSystem`. When I debug into the application, I can observe the file descriptors growing on my machine as well. It's just that I seem to get lucky that the garbage collector kicks in early enough before even more file descriptors are opened. When 
I increased the file descriptor limit on the test server, it finish running liquibase, but it only ran the finalizer and closed the file descriptor several minutes later (peak was ~6500 open file descriptors)

![image](https://user-images.githubusercontent.com/2456413/116455605-8fa23b00-a861-11eb-9bb9-1eebbaf8a762.png)

- - -
Is that enough information to reproduce? Not sure what the best solution here might be, but I guess some things that could be done are (i) closing the created file systems explicitly, (ii) not creating a new fallback accessor each time, (iii) not using the fallback accessor when the resource accessor from the current scope already returns a result.

**Comment by blutorange** [comment]
Also related to #1754 , where opening the JARs again and again also seems to slow down the entire process.

**Comment by vesin** [comment]
Hi @blutorange , what is the status of this issue?

**Comment by blutorange** [comment]
@vesin Unless something has been done on this (looking at release 4.3.4 and 4.3.5 it seems not), this issue should still exist.

To summarize:

* Use LiquiBase programmatically via `new Liquibase(changelog, new ClassLoaderResourceAccessor(...), connection).update(...)`
* We have many changelogs, on the order of 15 changelog files with 50-200 changesets.
* During the update process, as mentioned above, LiquibaseEntityResolver keeps creating new ClassLoaderResourceAccessors, which in turn keep creating more and more file handles to JAR files.
* Our application has ~300 JAR files, so each time 300 file descriptors are created.
* Currently Liquibase relies on the `finalize` method (of ZipFileSystem) to close the file descriptors. But depending on the garbage collector and the available RAM , this may not be called for several minutes. 
* This results in the system's limit being reached very quickly, so this is not usable in production.

As far as I can tell, this is not a random bug, but more like a conceptual issue with how  resource are handled, so I think somebody should definitely take a look at this.

So for the time being, we should be staying on 3.x until this gets fixed. Unfortunately, owasp-maven-plugin also reports critical security issues for 3.x, so we really need to update. We actually had to clone this project, make a quick and dirty fix to LiquibaseEntityResolver (just don't create a fallback resource accessor when there is a context resource accessor already) and compile that. But obiously that's not a good fix, a real fix would involve evaluating whether a fallback is needed, and if so, provide a single fallback that's not created everytime and close that at the end.

Also, we also had to subclass `ClassLoaderResourceAccessor` and add a `#close` method, since that class lazily opens file descriptors to JAR files, but provides not way of ever closing them, again relying on the `finalize` method.

**Comment by blutorange** [comment]
For reference, this might be the problematic commit: https://github.com/liquibase/liquibase/pull/1621 There it creates a new ClassLoaderResourceAccessor within the `resolve` method. Since then, creating the accessor was moved into an instance field, but this is still problematic since liquibase creates more than one `LiquibaseEntityResolver` instance when run.

**Comment by blutorange** [comment]
Also pretty similar to #1754 (probably the same cause), but that issue focuses on performance (doesn't seem to be slow for us), while this issue focuses on the application server process running out of resources

**Comment by nvoxland** [comment]
I updated the LiquibaseEntityResolver to be more efficient in how it creates ResourceAccessors and also made other code not construct as many instances of LiquibaseEntityResolver.

I was not able to replicate the original issue well to know for sure the fix solves the entire problem, though. Could you see if the fix in PR #1860 solves the problem for you, @blutorange ?

**Comment by blutorange** [comment]
@nvoxland  That sounds great! I'll take a look as soon as I can during the next days and will let you know.

**Comment by vesin** [comment]
I can also try beta version as soon as you can make it public.

**Comment by nvoxland** [comment]
You can get a build with the changes off https://github.com/liquibase/liquibase/actions?query=branch%3ALB-1547 

Can you try from that?

**Comment by Florianisme** [comment]
We have experienced similar performance issues as described in #1754 and the changes seem to have fixed it. I tested with the artifact linked by @nvoxland above.

**Comment by blutorange** [comment]
Sorry it took a bit longer, but I can only really test this at work and I was away for a few days. So first of all, the changes from https://github.com/liquibase/liquibase/tree/LB-1547 definitely fix the issue.

There's one minor request I would like to make. We're using liquibase as follows:

```java
var acc  = new ClassLoaderResourceAccessor(...);
try {
   var liqui = new Liquibase(file, acc, jdbcConnection);
   // ... do some work
}
finally {
   // not possible right now
   // acc.close();
}
```

What we'd like to do is close the resource accessor after use, so we can release the resources. Currently, ClassLoaderResourceAccessor has no close method -- could this be added? Otherwise it's never gets closed explicitly. It you let it implement `AutoCloseable` the implementation would look like this:

```java
public class ClassLoaderResourceAccessor extends AbstractResourceAccessor implements AutoCloseable {
  // ...

  @Override
  public void close() throws Exception {
    if (rootPaths != null) {
      for (final FileSystem rootPath : rootPaths) {
        try {
          rootPath.close();
        }
        catch (final Exception e) {
          // ignore, if it can't be closed it might be cleaned up by the system / GC later
        }
      }
    }
  }
}
```

**Comment by nvoxland** [comment]
Thanks for testing the fix.

Yes, adding a close method to the ResourceAccessor interface is a good idea. This ticket will likely go out in a 4.4.1 release and we don't want to add interface changes in a patch release, so I created #1873 to get into a main release.

**Comment by blutorange** [comment]
That sounds good, for now we can extend the class and add the close method ourselves. As for this ticket, I think we can close it then.

**Comment by vesin** [comment]
I tried solution with "AutoCloseable" and i still get same errors.
We have more then 300 change log files, and 40 tenants.

**Comment by blutorange** [comment]
@vesin You did test with the fixed version ([commit aeb34c9c91748926b207da18eeed14dca971b73a](https://github.com/liquibase/liquibase/commit/aeb34c9c91748926b207da18eeed14dca971b73a) it's not in Maven yet I believe?) What I did to analyze the problem was to debug which JARs are opened when. At least for me, I don't see any JARs being opened more than once at a time. Do you see a JAR being opened more than once at the same time? When you create a `ClassLoaderResourceAccessor`, it should open the JARs once, and then reuse that accessor in `LiquibaseENtityResolver` without creating any new once. Once liquibase is done, your code should close the `ClassLoaderResourceAccessor`, which in turn should close the opened JAR files.

For debugging I set breakpoints in `LiquibaseEntityResolver` and in the constructors of `jdk.nio.zipfs.JarFileSystem` / `jdk.nio.zipfs.ZipFileSystem`. You can also check how many files are opened via `ls /proc/<PID>/fd | wc -l` and get a list of opened files via `lsof -p 28537` (if you use Linux)

**Comment by vesin** [comment]
This is the code i use.

`@Service
public class LiquibaseService {
    private static final Logger LOG = LoggerFactory.getLogger(LiquibaseService.class);

    @Inject private MultiTenantConnectionProvider multiTenantConnectionProvider;

    public boolean initLiquibase(String tenant) throws LiquibaseException, SQLException {
        LOG.debug("Initializing liquibase");

        final ClosableClassLoaderResourceAccessor resourceAccessor = new ClosableClassLoaderResourceAccessor();
        try (final Connection connection = multiTenantConnectionProvider.getConnection(tenant);
                ) {
            LOG.debug("Liquibase connection acquired for {}", tenant + "_core");
            final Database database = DatabaseFactory.getInstance().findCorrectDatabaseImplementation(new JdbcConnection(connection));
            database.setDefaultSchemaName(tenant + "_core");
            final Liquibase liquibase = new Liquibase("liquibase/db-changelog.xml", resourceAccessor, database);
            if (liquibase.listLocks().length > 0) {
                LOG.debug("Liquibase already locked.");
                return false;
            }
            LOG.debug("Liquibase update started");
            liquibase.update(new Contexts(), new LabelExpression());
            liquibase.close();
            return true;
        }finally{
            resourceAccessor.close();
        }
    }
}

class ClosableClassLoaderResourceAccessor extends ClassLoaderResourceAccessor implements AutoCloseable {
    private static final Logger LOG = LoggerFactory.getLogger(ClosableClassLoaderResourceAccessor.class);
    @Override
    public void close() {
        if (rootPaths != null) {
            for (final FileSystem rootPath : rootPaths) {
                try {
                    LOG.info("Closing file" + rootPath.toString());
                    rootPath.close();
                } catch (final Exception e) {
                    LOG.error("Unable to close resource", e);
                    // ignore, if it can't be closed it might be cleaned up by the system / GC later
                }
            }
        }
    }
}`

**Comment by vesin** [comment]
I still can't understand why does liquibase needs to load all jars to execute change logs...

**Comment by nvoxland** [comment]
@vesin : Liquibase looks through all the jars because we support having changelog files packaged in a jar file, and so we need to check if they are in there. The base ClassLoader interface does have the functionality we need, so we need to open them ourselves to check separately. The fix should make it so they don't keep being re-opened though. If you aren't using a build with the fix from #1873 in it, even your auto-closable logic will probably not help. The bug was buried a bit deeper. 

For your use case, if you know better where your files will be you can use a different ResourceAccessor setup too. For example, if you use the `ClassLoaderResourceAccessor(ClassLoader)` constructor instead, it will look only in the locations specified in the passed classloader vs. all the jars in your entire application. There is also a FileSystemResourceAccessor that could work for you as well. A different classloader setup could avoid the bug enough to solve your problem until the actual fix goes out.

**Comment by vesin** [comment]
Great, thanks!

**Comment by stolp** [comment]
> Thanks for testing the fix.
> 
> Yes, adding a close method to the ResourceAccessor interface is a good idea. This ticket will likely go out in a 4.4.1 release and we don't want to add interface changes in a patch release, so I created #1873 to get into a main release.

I see that #1873 is not resolved in the 4.4.0 release. Does this mean we cannot expect a fix before 4.5.0?

**Comment by vesin** [comment]
Hi all, when can we expect this fix to be released?
